Rewrite to use sqlite3 and public api of publishing-api, ready for jenkins

benhyland opened the pull request:
# Goals of this PR

Use sqlite3 for our local db to avoid having to instal postgres on the
Jenkins node which will run the checker.

Use a new endpoint on Publishing API so that we are using a maintained, public API.
This avoids the fragility and coupling of using the publishing-api database directly.

Define the checks we wish to run in order to detect interesting discrepancies
between Rummager and Pubilshing API content. Report on the results with enough detail
to allow for useful analysis by humans.

Redesign the program so that is is suitable for Jenkins, and can be more easily maintained in future.
In particular, introduce a whitelist so that we can keep Jenkins green while we work on fixing
any problems found.
# Still to come

Add rubocop/linting.

Analyse the check outputs and add appropriate whitelisting.

Use the checker in Jenkins, once https://github.com/alphagov/publishing-api/pull/305 is merged.
In the meantime, the checker can be run locally by pointing it at a publishing-api from that PR branch.
Therefore, this PR can already replace the previous version.
# Review

This PR is the result of many small commits, which were unfortunately mostly not addressing particular areas of the program in an orderly manner but rather exploring the entire program piecemeal.
It would be rather onerous and conflict-ridden to attempt reordering. Commits have been squashed into a few large chunks, which might represent the failings and increased understanding of the authors over time; however, reviewers will probably find it easier consider the overall result rather than the individual commits.

See https://trello.com/c/DYFgXNYr/621-automate-link-checker-tool

@MatMoore @Davidslv @benhyland

Commit:
Start on rewriting for sqlite and publishing-api

We are moving to sqlite for our local db to avoid a dependency on postgres
on the jenkins node which will run the checker.

We are introducing a new endpoint for publishing-api to allow
us to query public apis rather than go to the publishing-api db.

This is the first pass of the necessary rewrite, establishing the skeleton for import
and checks, and getting the data import working with sqlite as it was before. We have changed
the imported table structure to better suit our needs and because sqlite3 does not support
the same data types as postgres.

We use gds-api-adapters where possible, and we will need to switch to using it for the
new endpoint when that endpoint is finished.

MatMoore commented on lib/checks/require/all_checks.rb:
```diff
@@ -0,0 +1,10 @@
+module Checks
+  class Require
+
+    private_class_method :initialize
```
I think here you can just make `Require` a module instead of a class, rather than make `initialize` private.

i.e.

``` ruby
module Checks
  module Require
    def self.all_checks
    end
  end
end
```

jackscotti commented:
_"Always leave a trailing newline in the file to keep compatibility with command-line tools. Some text editors will add this automatically; Sublime Text will add it if you enable the ensure_newline_at_eof_on_save setting. (TrailingBlankLines)"_

https://github.com/alphagov/styleguides/blob/master/ruby.md#general

MatMoore commented on lib/import/publishing_api_importer.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+
+    def import
+      create_rummager_tables
```
think this should be create_publishing_api_tables

MatMoore commented:
Looks good so far. I was wondering if it's possible/sensible to use a test framework for check discovery/execution so we can report results in a more standard way?

benhyland commented:
@MatMoore Discussed using a test framework to represent our checks offline. We decided that the value gained is not worth the effort in this particular example, since to do this we would want to further separate the environment and import from the tests, and this is difficult as we are using in-memory sqlite.

Davidslv commented on lib/check_runner.rb:
```diff
@@ -0,0 +1,40 @@
+private
+
+  def run_importers
+    @importers.map { |importer| Thread.new { importer.import } }.each { |thread| thread.join }
```
you can do `.each(&:join)` saves you writing another block

Davidslv commented on lib/import/rummager_data_presenter.rb:
```diff
@@ -0,0 +1,52 @@
+  private
+
+    def self.encode_array(resource)
+      "{#{resource.join(',')}}"
```
we might not need this anymore as this was to mimic postgresql array representation, SQLite doesn't support arrays

Davidslv commented on lib/import/publishing_api_importer.rb:
```diff
@@ -0,0 +1,56 @@
+module Import
+  class PublishingAPIImporter
```
nit picking but `PublishingAPIImporter` should be `PublishingApiImporter`

Davidslv commented on spec/import/rummager_data_presenter_spec.rb:
```diff
@@ -0,0 +1,44 @@
+      row_data = test_data_example
+
+      expected_rows = [
+          ["/vehicle-tax", "organisations", "{/government/organisations/driver-and-vehicle-licensing-agency,/government/organisations/test-org}"],
```
those curly braces represent an array in postgresql, should we keep it or simply remove them and comma separate?

benhyland replied:
We'll most likely end up changing this but I didn't want to think about it until we are doing the comparison checks.

Commit:
WIP for writing the checks

Improve the data importing - speed and structure.
Add exit code based on the check outcomes, for jenkins to use.

Commit:
Write all the checks we think we need

We provide additional data in the import to support human analysis
and whitelisting of check output.

Commit:
Add basic whitelisting functionality

We will need to suppress output from checks that we know about in order to keep
Jenkins green. For example, something we know is being fixed already, or something
for which we need to delay the fix.

If we decide to *never* fix something, we should filter it in the check rather than in
the whitelist. Therefore, every whitelist entry can and should have an expiry date and
a reason for existing. This allows us to report errors if any whitelist entry is expired.

malcolmbaig commented on lib/checks/base_paths_missing_from_publishing_api.rb:
```diff
@@ -0,0 +1,31 @@
+      AND format NOT IN ('recommended-link')
+      SQL
+
+      name = self.class.name.split('::').last
```
Seem to be doing this a lot - wonder if it's cleaner just passing `self.class` into `Report` and doing the name extracting logic in there.

benhyland replied:
I didn't bother before because the value is needed for the whitelist also. Really it's a property of the check, which should be extracted by the common check creation code. I'll address it there.

malcolmbaig commented on lib/checks/links_missing_from_publishing_api.rb:
```diff
@@ -0,0 +1,45 @@
+
+      name = self.class.name.split('::').last
+      headers = ['link_type', 'link_content_id', 'content_id', 'publishing_app', 'format']
+      publishing_api_missing_links = @whitelist.apply(name, headers, @checker_db.execute(publishing_api_missing_links_query))
```
can we multi-line this for readability

benhyland replied:
sure thing

malcolmbaig commented on lib/checks/base_paths_missing_from_publishing_api.rb:
```diff
@@ -0,0 +1,31 @@
+    def run_check
+      query = <<-SQL
+      SELECT
+      rc.base_path,
```
A general style observation - I find it's a lot easier to read and understand these queries when these abbreviations are fully expanded into their respective table names. Saves having to remember what they are / jump back to the line containing the alias.

benhyland replied:
Just my previous habit. I'll fix, and we can avoid aliasing in future if it helps.

benhyland replied:
I thought this was fixable but on second thoughts I find the non-aliased version much harder to read wherever the table name is long.
Since readability is subjective I'm going to leave this as is, we can discuss offline if it really horrifies you.
In general we should trust the pair and be consistent with the existing file.

malcolmbaig commented on lib/checks/base_paths_missing_from_rummager.rb:
```diff
@@ -0,0 +1,34 @@
+    def run_check
+      query = <<-SQL
+      SELECT
+      pac.content_id,
```
Indentation in these select statements is inconsistent with other `Check` classes. Let's pick one style and apply throughout. FWIW, I think indenting reads easier.

benhyland replied:
I'll indent select within sql heredoc strings everywhere.

malcolmbaig commented on lib/checks/expired_whitelist_entries.rb:
```diff
@@ -0,0 +1,20 @@
+    # find whitelist entries which have expired
+
+    def run_check
+
```
whitespace

benhyland replied:
can you be more specific? Do we have particular linting rules you are manually applying? If so I suggest we wait until lint and rubocop are added to this repo (the task is already on our list).

malcolmbaig commented on lib/db/checker_db.rb:
```diff
@@ -0,0 +1,65 @@
+    # SQLITE_LIMIT_COMPOUND_SELECT defaults to 500
+    rows.each_slice(500) do |slice_rows|
+      insert_sqlite_batch(
+          table_name: table_name,
```
indent's a bit far out

benhyland replied:
Fixing; my continuation indent setting was wrong.

malcolmbaig commented on lib/import/publishing_api_importer.rb:
```diff
@@ -0,0 +1,100 @@
+
+    def create_publishing_api_tables
+      @checker_db.create_table(
+          table_name: 'publishing_api_content',
```
indent checks needed in this method

benhyland replied:
yep, bad IDE config as above

malcolmbaig commented on lib/import/publishing_api_data_presenter.rb:
```diff
@@ -0,0 +1,37 @@
+    end
+
+    def self.present_links(content_id_data)
+
```
whitespace

benhyland replied:
as above, waiting for automatic lint

malcolmbaig commented on lib/import/publishing_api_data_presenter.rb:
```diff
@@ -0,0 +1,37 @@
+    end
+
+    def self.ever_published(content_id_data)
+      published_state = content_id_data['content_items'].any? { |ci| ['published', 'unpublished', 'superseded'].include?(ci['state']) }
```
multiline for readability:

```
published_state = content_id_data['content_items'].any? do |ci| 
  ['published', 'unpublished', 'superseded'].include?(ci['state']) 
end
```

benhyland replied:
pretty subjective, but OK

Commit:
Address review comments

- pass name in to checks to avoid duplicated introspection
- indent selected cols within sql heredoc strings
- some line breaks for readability
- indent correction (the continuation indent on my editor was set wrongly)
- add comments

malcolmbaig commented on lib/import/publishing_api_importer.rb:
```diff
@@ -0,0 +1,100 @@
+
+    def initialize(checker_db, progress_reporter, publishing_api_url)
+      @checker_db = checker_db
+      @thread_pool = Thread.pool(1)
```
How come this is 1 vs the 3 used in the Rummager importer?

benhyland replied:
The pools in each importer are for additional work that needs to be performed when an api response arrives.
The main api iteration for each importer happens on separate threads owned by the CheckRunner. The two importers have different work to perform when a response comes in - in particular, the Rummager importer actually calls out to publishing api. As a result fewer auxiliary threads are needed by the Publishing API importer.
This could all be much more elegant but we've reached (and passed) the point of diminishing returns IMO.
Also note that the pools are sized for a dev VM - presumably we could hammer the apis harder from Jenkins in integration but there's not much need since the import takes ~15mins.

malcolmbaig commented on lib/import/publishing_api_importer.rb:
```diff
@@ -0,0 +1,100 @@
+      response = do_request(NIL_UUID)
+      results = response["results"]
+
+      # todo: can we get this for real? needs a new endpoint... probably not worth it. just supply a different progress reporter?
```
Suggested rewording: `#TODO: either change the progress reporter so this hard-coded value isn't needed, or determine how to retrieve the value from the publishing-api`.

Also, please format to >80 char.

MatMoore replied:
I don't think there's a need for a new endpoint; we can just run a separate count query as part of the pagination on the publishing api side. It can change as part of https://trello.com/c/3qIYFDBI/634-finish-publishing-api-endpoint-for-link-checker-tool

OTOH I'm fine with just hardcoding a guess for now as the progress report is just for giving the user some feedback.

benhyland replied:
I'll adjust the comment and remove the TODO.
(as an aside, line length should be enforced by lint or not at all).

malcolmbaig commented on lib/import/publishing_api_importer.rb:
```diff
@@ -0,0 +1,100 @@
+          rows: rows
+        )
+    end
+
```
whitespace

benhyland replied:
as above, will handle via linting

malcolmbaig commented on lib/import/rummager_data_presenter.rb:
```diff
@@ -0,0 +1,47 @@
+    end
+
+    def self.present_links(row_data)
+
```
whitespace

benhyland replied:
as above, will handle via linting

malcolmbaig commented on lib/import/rummager_data_presenter.rb:
```diff
@@ -0,0 +1,47 @@
+  module RummagerDataPresenter
+
+    def self.present_content(batch_data)
+      batch_data.map { |item| [item['link'], item['content_id'], item['format'], item['index'], item['document_type']] }
```
suggest 

```
batch_data.map do |item| 
  [
    item['link'], 
    item['content_id'], 
    item['format'], 
    item['index'], 
    item['document_type']
  ] 
end
```

for readability

benhyland replied:
yep, fixed

malcolmbaig commented on lib/import/rummager_data_presenter.rb:
```diff
@@ -0,0 +1,47 @@
+      mainstream_browse_pages = fetch_resources(row_data, "mainstream_browse_pages", "/browse/")
+
+      rows_for(base_path, 'policies', policies) +
+          rows_for(base_path, 'people', people) +
```
indentation

benhyland replied:
as above, will handle via linting

malcolmbaig commented on lib/import/rummager_data_presenter.rb:
```diff
@@ -0,0 +1,47 @@
+      return nil unless resources
+
+      resources.map { |resource|
+          case resource
```
indentation

benhyland replied:
fixed

malcolmbaig commented on lib/import/rummager_data_presenter.rb:
```diff
@@ -0,0 +1,47 @@
+      resources = row_data[resource_name]
+      return nil unless resources
+
+      resources.map { |resource|
```
please use `do..end` instead of `{..}`

benhyland replied:
That's a Ruby convention for multiline vs single line blocks? Sure.

malcolmbaig commented on lib/import/rummager_importer.rb:
```diff
@@ -0,0 +1,148 @@
+      @progress_reporter = progress_reporter
+
+      @rummager = GdsApi::Rummager.new(
+          Plek.new.find('rummager'),
```
indentation

benhyland replied:
fixed

malcolmbaig commented on lib/import/rummager_importer.rb:
```diff
@@ -0,0 +1,148 @@
+          timeout: 20,
+      )
+      @publishing_api = GdsApi::PublishingApiV2.new(
+          Plek.new.find('publishing-api'),
```
indentation

benhyland replied:
fixed

malcolmbaig commented on lib/import/rummager_importer.rb:
```diff
@@ -0,0 +1,148 @@
+      create_rummager_tables
+
+      import_rummager_batches do |batch_data|
+
```
whitespace

benhyland replied:
as above, will handle via linting

malcolmbaig commented:
Got a little bit left to review on this, will continue tomorrow when Github isn't quite so broken.

Rather than be annoying with my whitespace/indent comments, just a general observation that we could do with a quick run-through of all changed files to remove extra newlines and correct formatting to 2-space indents.

General approach looks fine to me so far, just the odd question here and there.

MatMoore commented on lib/import/publishing_api_importer.rb:
```diff
@@ -0,0 +1,100 @@
+      end
+    end
+
+    def do_request(last_seen_content_id)
```
We'll change this to use gds_api_adapters soon. Doesn't haven't to block this PR though.

benhyland replied:
yay :)

MatMoore commented on lib/whitelist/whitelist.rb:
```diff
@@ -0,0 +1,43 @@
+  end
+
+  def get_or_else(hash, key, default)
+    # can't use #fetch since it returns nil if the key exists but maps to nil
```
can you explain this? Fetch can take a default value to instead of nil if that was the problem?

`hash.fetch(key, default)`

benhyland replied:
I believe the default you can provide to `fetch` is used when the key is missing, but not when it is present and maps to the value `nil`.
'Billion dollar mistake' was a large underestimate :P

MatMoore commented on lib/whitelist/whitelist.rb:
```diff
@@ -0,0 +1,43 @@
+    conjunctions = conj_hash_arr.map { |conj_hash| Whitelist.conjunction_for(headers, conj_hash) }
+    lambda { |row| conjunctions.any? { |c| c.call(row) } }
+  end
+
```
Can you define the whitelist format and semantics in a comment somewhere in this file?

I think defining the terms (predicate, conjunction etc) will help a lot.

benhyland replied:
I added a big comment and linked to https://en.wikipedia.org/wiki/Disjunctive_normal_form
Please lmk if it needs more attention.

MatMoore commented on lib/whitelist/whitelist.rb:
```diff
@@ -0,0 +1,43 @@
+  end
+
+  def report_expired_entries(reference_date)
+    expiries = @whitelist.flat_map { |check| get_or_else(check[1], 'rules', []).map { |rule| [check[0], rule['expiry'], rule['reason']]} }
```
I find it hard to scan long lines like this because you have to pick out the method calls before it makes sense.

When I first read this I didn't even realise the second map was part of the first block because I saw some dots and methods and just assumed it was method chaining.

It might help if you extracted a method from this, but I think splitting over multiple lines would make it a lot clearer.

Something like

``` ruby
@whitelist.flat_map do |check_name, check|
  check.fetch('rules', []).map do |rule|
    [rule_name, rule['expiry'], rule['reason']]
  end
end
```

benhyland replied:
Sure. I guess everyone has different syntax highlighting and tolerance for this sort of nonsense :)

benhyland commented:
Thanks all for comments so far and for persistence in the teeth of github being on fire.
I believe I've addressed everything so far in line notes.
I have pushed but GH is still a smoking ruin, so may take a while to appear.

tijmenb commented:
I'm going to merge this so any other comments can be addressed in smaller PRs. 👍
